pkg/inclusion/nmt_caching.go

```diff
@@ -0,0 +1,123 @@
+// walk throws an error if the sub tree cannot be found.
+func (strc subTreeRootCacher) walk(root []byte, path []bool) ([]byte, error) {
+	// return if we've reached the end of the path
+	if len(path) == 0 {
+		return root, nil
+	}
+	// try to lookup the provided sub root
+	children, has := strc.cache[string(root)]
+	if !has {
+		// note: we might want to consider panicing here
+		return nil, fmt.Errorf("did not find sub tree root: %v", root)
+	}
+
+	// continue to traverse the tree by recursively calling this function on the next root
+	switch path[0] {
+	// walk left
+	case false:
+		return strc.walk([]byte(children[0]), path[1:])
+	// walk right
+	case true:
+		return strc.walk([]byte(children[1]), path[1:])
+	default:
+		// this is unreachable code, but the compiler doesn't recognize this somehow
+		panic("bool other than true or false, computers were a mistake, everything is a lie, math is fake.")
+	}
+}
+
+// EDSSubTreeRootCacher caches the inner nodes for each row so that we can
+// traverse it later to check for message inclusion. NOTE: Currently this has to
+// use a leaky abstraction (see docs on counter field below), and is not
+// threadsafe, but with a future refactor, we could simply read from rsmt2d and
+// not use the tree constructor which would fix both of these issues.
+type EDSSubTreeRootCacher struct {
+	caches     []*subTreeRootCacher
+	squareSize uint64
+	// counter is used to ignore columns NOTE: this is a leaky abstraction that
+	// we make because rsmt2d is used to generate the roots for us, so we have
+	// to assume that it will generate a row root every other tree contructed.
+	// This is also one of the reasons this implementation is not thread safe.
+	// Please see note above on a better refactor.
+	counter int
+}
+
+func NewCachedSubtreeCacher(squareSize uint64) *EDSSubTreeRootCacher {
+	return &EDSSubTreeRootCacher{
+		caches:     []*subTreeRootCacher{},
+		squareSize: squareSize,
+	}
+}
+
+// Constructor fullfills the rsmt2d.TreeCreatorFn by keeping a pointer to the
+// cache and embedding it as a nmt.NodeVisitor into a new wrapped nmt.
+func (stc *EDSSubTreeRootCacher) Constructor() rsmt2d.Tree {
+	// see docs of counter field for more
+	// info. if the counter is even or == 0, then we make the assumption that we
+	// are creating a tree for a row
+	var newTree wrapper.ErasuredNamespacedMerkleTree
+	switch stc.counter % 2 {
+	case 0:
+		strc := newSubTreeRootCacher()
+		stc.caches = append(stc.caches, strc)
+		newTree = wrapper.NewErasuredNamespacedMerkleTree(stc.squareSize, nmt.NodeVisitor(strc.Visit))
+	default:
+		newTree = wrapper.NewErasuredNamespacedMerkleTree(stc.squareSize)
+	}
+
+	stc.counter++
+	return &newTree
+}
+
+// GetSubTreeRoot traverses the nmt of the selected row and returns the
+// subtree root. An error is thrown if the subtree cannot be found.
+func (stc *EDSSubTreeRootCacher) GetSubTreeRoot(dah da.DataAvailabilityHeader, row int, path []bool) ([]byte, error) {
```

Comment:
there is a check on rows, its just kinda hidden

```go
	if len(stc.caches) != len(dah.RowsRoots) {
		return nil, fmt.Errorf("data availability header has unexpected number of row roots: expected %d got %d", len(stc.caches), len(dah.RowsRoots))
	}
	if row >= len(stc.caches) {
		return nil, fmt.Errorf("row exceeds range of cache: max %d got %d", len(stc.caches), row)
	}
```

only relying on the error from walking to do the check instead of checking for max length. There probably could be, but since we're the only ones using this functions, it seems a bit overkill to compute the square root of the length of the dah roots

we might be able to not export this entire struct actually, come to think of it. will wait until we add the get commitment function used in the implement. If we're not using any of these functions we can unexport them